src/livestatus.c

```diff
@@ -0,0 +1,463 @@
+  "log_messages_rate "                                                         \
+  "neb_callbacks "                                                             \
+  "neb_callbacks_rate "                                                        \
+  "requests "                                                                  \
+  "requests_rate "                                                             \
+  "service_checks "                                                            \
+  "service_checks_rate"
+
+#define LIVESTATUS_EXPECTED_FIELDS_RESP_NB 19
+
+struct livestatus_status_s {
+  int cached_log_messages;
+  lcounter_t connections;
+  double connections_rate;
+  lcounter_t forks;
+  double forks_rate;
+  lcounter_t host_checks;
+  double host_checks_rate;
+  lcounter_t livecheck_overflows;
+  double livecheck_overflows_rate;
+  lcounter_t livechecks;
+  double livechecks_rate;
+  lcounter_t log_messages;
+  double log_messages_rate;
+  lcounter_t neb_callbacks;
+  double neb_callbacks_rate;
+  lcounter_t requests;
+  double requests_rate;
+  lcounter_t service_checks;
+  double service_checks_rate;
+};
+typedef struct livestatus_status_s livestatus_status_t;
+
+static livestatus_t livestatus_obj;
+
+livestatus_status_t c_to_livestatus_status(const char **fields) {
+  livestatus_status_t status = {
+      .cached_log_messages = atoi(fields[0]),
+      .connections = strtoull(fields[1], NULL, 10),
+      .connections_rate = strtod(fields[2], NULL),
+      .forks = strtoull(fields[3], NULL, 10),
+      .forks_rate = strtod(fields[4], NULL),
+      .host_checks = strtoull(fields[5], NULL, 10),
+      .host_checks_rate = strtod(fields[6], NULL),
+      .livecheck_overflows = strtoull(fields[7], NULL, 10),
+      .livecheck_overflows_rate = strtod(fields[8], NULL),
+      .livechecks = strtoull(fields[9], NULL, 10),
+      .livechecks_rate = strtod(fields[10], NULL),
+      .log_messages = strtoull(fields[11], NULL, 10),
+      .log_messages_rate = strtod(fields[12], NULL),
+      .neb_callbacks = strtoull(fields[13], NULL, 10),
+      .neb_callbacks_rate = strtod(fields[14], NULL),
+      .requests = strtoull(fields[15], NULL, 10),
+      .requests_rate = strtod(fields[16], NULL),
+      .service_checks = strtoull(fields[17], NULL, 10),
+      .service_checks_rate = strtod(fields[18], NULL),
+  };
+
+  return status;
+} /* livestatus_status_t c_to_livestatus_status */
+
+static int ls_init(void) {
+  livestatus_obj.socket_file = "/var/cache/naemon/live";
+  livestatus_obj.max_retry = 20;
+  livestatus_obj.backoff_sec = 1;
+
+  return 0;
+} /* int ls_init */
+
+static int ls_config(const char *key, const char *value) /* {{{ */
+{
+  if (strcasecmp("LivestatusSocketFile", key) == 0) {
+    char *socket_file = strdup(value);
+    if (socket_file == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return 1;
+    }
+
+    livestatus_obj.socket_file = socket_file;
```

Comment:
Nice catch. How would you like me to handle this ? When should I `free` the value ?

I've looked at the _zookeeper_ plugin and `sfree()` is called in the _config callback_.
Such approach does not free the value if the plugin is unregistered, am I right ?
Thanks in advance for your suggestions.

Reply:
Note that, if `.socket_file` is initialized to a string literal, freeing the literal will segfault. What we typically do instead is initialize the field to `NULL` and, on use, do something like:

```c
(obj.socket_file != NULL) ? obj.socket_file : DEFAULT_SOCKET_FILE
```

Reply:
Hi @octo . I've hooked into the `shutdown` callback to free the memory (very much inspired from the `chrony` plugin).
Question: Since _collectd_ can dynamically unregister plugins, maybe that relying on the `shutdown` callback is not enough. What's your advice ?